flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/AbstractAsyncStateStreamOperator.java

```diff
@@ -330,7 +332,22 @@ public void processWatermark(Watermark mark) throws Exception {
             super.processWatermark(mark);
             return;
         }
-        asyncExecutionController.processNonRecord(() -> super.processWatermark(mark));
+        asyncExecutionController.processNonRecord(
+                () -> {
+                    // todo: make async operator deal with interruptible watermark
+                    if (timeServiceManager != null) {
+                        CompletableFuture<Void> future = timeServiceManager.advanceWatermark(mark);
+                        if (asyncExecutionController.isSerialMode()) {
+                            asyncExecutionController.drainInflightRecords(0);
+                            future.get();
+                            output.emitWatermark(mark);
+                        } else {
+                            future.thenAccept(v -> output.emitWatermark(mark));
+                        }
```

Comment:
How about package these logic into a method of `AEC`?
e.g.

```suggestion
                        CompletableFuture<Void> future = timeServiceManager.advanceWatermark(mark);
                        future.thenAccept(v -> output.emitWatermark(mark));
                        asyncExecutionController.drainWithTimer();
```

And we only do `assert future.isDone()` in the `drainWithTimer()` instead of the `future.get()`

Reply:
And I think we could handle the interruptible timers in later PR(s).

Reply:
Nice suggestion, moved the `drain` and `assert future.isDone()` into AEC.